### ISSUE TITLE
Hybrid Quantum Algorithms Benchmarking

### ISSUE THREAD

**Issue by ceasar88** [SURFQuantum/QML-Medicine#11]
https://github.com/SURFQuantum/qc-hybrid-benchmarking/tree/split_benchmarking_and_profiling

🎯 Objective Summary

The primary objective of the framework is to provide a standardized, modular, and quantitative method for comparing the efficiency and efficacy of different hybrid quantum approaches against each other and against their classical counterparts.

**Comment by ceasar88** [comment]
** AI GENERATED **

<h1 data-pm-slice="1 3 []">The Hybrid Benchmarking Framework</h1><h2>🎯 Overview</h2><p>The <strong>Hybrid Benchmarking Framework</strong> (<code>qc-hybrid-benchmarking</code>) is a composition-based Python package designed to measure and compare the performance of hybrid classical-quantum algorithms. Its structure is modular, allowing new metrics and visualization tools to be easily added.</p><h2>I. Core Components</h2><p>These files form the foundation and execution engine of the framework.</p><h3>1. <code>benchmark_framework.py</code></h3><p>This is the central execution engine. It initializes the run, manages the context, calls the external adapter, collects the raw results, and passes them to all registered metrics for analysis.</p>
Role | Description
-- | --
Executor | Runs the external algorithm via the adapter function.
Compositor | Manages the list of active metrics.
Reporter | Generates the final console and plot reports.

<p><strong>Code Example (Base Class):</strong></p><pre><code>from abc import ABC, abstractmethod
from typing import Dict, Any

class Metric(ABC):
    """Abstract base class for all benchmark metrics."""
    
    @abstractmethod
    def process_results(self, raw_results: Dict[str, Any]):
        """Calculates metric value from raw data."""
        pass
    
    @abstractmethod
    def generate_report_line(self) -&gt; Dict[str, str]:
        """Returns the key-value pair for the final report."""
        pass
</code></pre><h2>II. Specific Metrics</h2><p>These are concrete implementations of the <code>Metric</code> base class, each focusing on a different aspect of performance.</p><h3>1. <code>metrics/optimality.py</code></h3><p>Measures the quality of the final result relative to an expected minimum.</p><ul><li><p><strong>Key Measurement:</strong> <strong>Optimality Gap</strong>, calculated as the difference between the final loss value and the <code>known_minimum</code> defined in the context.</p></li></ul><h3>2. <code>metrics/profiling.py</code></h3><p>Measures the efficiency and resource consumption, focusing heavily on timing.</p><ul><li><p><strong>Key Measurement:</strong> <strong>Time Profiler</strong>. Calculates the total run time and, critically, aggregates the <strong>Total Quantum Time</strong> by summing the execution times from the <code>quantum_eval_times</code> list provided by the adapter.</p></li></ul><h3>3. <code>metrics/depth.py</code></h3><p>Analyzes the structural complexity of the quantum circuits.</p><ul><li><p><strong>Key Measurement:</strong> <strong>Circuit Depth</strong> and <strong>Gate Count</strong>, derived from the circuit description provided in the configuration.</p></li></ul><h2>III. Internal Utilities</h2><h3>1. <code>visualization/visualizer.py</code></h3><p>Handles the generation of graphical reports.</p><ul><li><p><strong>Role:</strong> Takes the <code>cost_history</code> from the raw results and generates standard plots (e.g., loss vs. epoch, time breakdown charts), saving them to a dedicated directory (<code>benchmark_plots/</code>).</p></li></ul>

**Comment by ceasar88** [comment]
** AI GENERATED **

<h1 data-pm-slice="1 3 []">🔬 Integration and Benchmark Report: QCNN Model (8-Qubit)</h1>

Code: https://github.com/SURFQuantum/QML-Medicine/tree/benchmark_framework

<h2>🎯 Executive Summary</h2><p>The <strong>Hybrid Benchmarking Framework</strong> has been successfully integrated into the <code>QML-Medicine</code> repository. The 8-qubit quantum mode of the QCNN model was executed using the framework's time-profiling and metric gathering capabilities.</p><p><strong>Key Findings:</strong> The model successfully trained for 50 iterations, achieving a final cost of approximately <strong>2.86</strong>. The entire training process took <strong>91.85 seconds</strong>. The total reported quantum time was 0.0s, indicating an issue with the instrumentation or the quick execution of the simulation device, requiring further investigation.</p><h2>I. Integration Status</h2><p>The integration involved setting up a clear interface between the QCNN training code and the external framework.</p><ul><li><p><strong>Adapter (<code>qcnn_wrapper.py</code>):</strong> This script successfully instrumented the PennyLane device to capture individual QNode execution times, although the resulting sum was reported as 0.0s. It handled the necessary data formatting to feed the training results into the framework.</p></li><li><p><strong>Executor (<code>run_benchmark.py</code>):</strong> This script successfully loaded the configuration and composed the framework with the <code>OptimalityMetric</code>, <code>TimeProfiler</code>, and <code>CircuitDepthMetric</code> to run the job.</p></li><li><p><strong>Data Integrity:</strong> All prior module and data-length issues (Pandas indexing errors) were resolved, allowing the full execution to complete and generate the raw results dictionary.</p></li></ul><h2>II. Benchmark Run Details</h2><h3>Test Run Configuration</h3>
Parameter | Value
-- | --
Run Label | QCNN_Benchmark_quantum_8q
Model | Hybrid QCNN
Execution Mode | Quantum (Simulated)
Number of Qubits | 8
Total Iterations | 50
Known Minimum | $0.1$

<h2>III. Optimization Cost History</h2><p>The plot below illustrates the model's convergence over the 50 iterations. The cost drops quickly in the first 20 iterations before plateauing near the final value.</p><p>The red dashed line represents the <code>Known Minimum Energy</code> (<math-inline data-math="0.1">$0.1$</math-inline>), showing that the model converged far from the expected optimal value, indicating a potential need for hyperparameter tuning or a deeper circuit.</p><h2>

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/4590075c-48f6-415d-ae7b-a4d26d079f73" />

IV. Analysis and Next Steps</h2><ol><li><p><strong>Quantum Time Investigation:</strong> The <strong><code>Total Quantum Time</code> reported as 0.0 seconds</strong> is highly unusual for a 50-epoch run. This suggests one of two scenarios:</p><ul><li><p>The PennyLane device instrumentation hook was not active or correctly capturing the time.</p></li><li><p>The simulation device (e.g., <code>default.qubit</code> on a fast CPU) executed each QNode call too quickly for Python's <code>time.monotonic()</code> to register a measurable difference.
<strong>Action:</strong> Review the adapter's time-capturing logic to ensure precision.</p></li></ul></li><li><p><strong>Efficacy:</strong> The model's final cost is high (<math-inline data-math="\approx 2.86">$\approx 2.86$</math-inline>).
<strong>Action:</strong> Proceed with hyperparameter search (learning rate, batch size) or explore a deeper QCNN architecture before comparing to the classical baseline.</p></li></ol>
